sdks/python/apache_beam/runners/dataflow/internal/apiclient.py

```diff
@@ -740,6 +719,25 @@ def _apply_sdk_environment_overrides(
       new_payload.container_image = new_container_image
       environment.payload = new_payload.SerializeToString()
 
+    # De-dup environments by Docker container image since currently Dataflow
```

Comment:
Thanks for tagging me on this change. I am working on a related change to reflect pipeline resource hints in portable proto. Hints are defined in `Environment.resource_hints`. Transforms are mapped to environments, and different transforms can have different hints. Therefore, we can have multiple environments with different hints, but the same container image.
My change is not yet ready to review, but the replication logic looks like this:
https://github.com/apache/beam/pull/14082/files#diff-252b68d1b24f6f7cdd8c5e54163d4856afad59fd385f5f6a91bf0fe66f09e67dR243 

I think deduplicating logic as proposed in this change will be difficult to reconcile with resource hints representation.

Reply:
To clarify, the restriction is already there for Dataflow. We currently start an SDK Harness per container image an de-dup here: https://github.com/apache/beam/blob/83bd5485047373ae0e380c54063e3769874a8b09/sdks/python/apache_beam/runners/dataflow/internal/apiclient.py#L304

This just moves the de-duping from container images started to Dataflow to environments in the proto since I'm trying to update Dataflow to map work items to environments based on the environment ID (not container image).

I can try to reduce de-duping to multi-language Java environments in muti-language pipelines since multiple Python environments do not seem to be running into issues currently. Multiple Java environments in multi-language pipelines run into dependency conflicts. Does that help ?

Reply:
Restricted de-duping to multi-language Java environments.